### PR TITLE
ref(types): Add types to Relay endpoints

### PR DESCRIPTION
**Pull request by mgaeta**
Endpoint files tend not to have a lot of functions to annotate so this is mostly just to mark the directory moving forward.

### CODE REVIEW

File: src/sentry/api/endpoints/relay/health_check.py [hunk 1]
```diff
@@ -14,8 +14,8 @@ class RelayHealthCheck(Endpoint):
     Relay doesn't need to care if it connects to another Relay or directly to sentry.
     """
 
-    authentication_classes = []
-    permission_classes = []
+    authentication_classes = []  # type: ignore
```
Comment (evanh) [inline src/sentry/api/endpoints/relay/health_check.py:17]: Why are these ignored?

File: src/sentry/api/endpoints/relay/project_configs.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import logging
 import random
+from typing import List, MutableMapping, Set
```
Comment (evanh) [inline src/sentry/api/endpoints/relay/project_configs.py:3]: If you add `from __future__ import annotations` here you can type things with the builtin `set` and `list` instead of `Set` and `List` respectively.

File: src/sentry/api/endpoints/relay/register_challenge.py [hunk 3]
```diff
@@ -50,7 +50,7 @@ def post(self, request: Request) -> Response:
 
         public_key = json_data.get("public_key")
         if not public_key:
-            return Response({"detail": "Missing public key"}, status=status.HTTP_400_FORBIDDEN)
+            return Response({"detail": "Missing public key"}, status=status.HTTP_400_BAD_REQUEST)
```
Comment (evanh) [inline src/sentry/api/endpoints/relay/register_challenge.py:53]: Why this change?

Comment (evanh) [reply src/sentry/api/endpoints/relay/register_challenge.py:53]: Sorry, are you changing this to return a 400 instead of a 500?

Comment (mgaeta) [reply src/sentry/api/endpoints/relay/register_challenge.py:53]: In order to satisfy mypy, I can either do:
```py
return Response({"detail": "Missing public key"}, status=status.HTTP_400_BAD_REQUEST)
```
which would fix the bug and start returning a 400, or 
```py
return Response({"detail": "Missing public key"}, status=status.HTTP_400_FORBIDDEN)  #type: ignore
```
which will leave the bug in place and keep returning 500s.

File: src/sentry/api/endpoints/relay/health_check.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/api/endpoints/relay/health_check.py:17]: These variables override their parents so they're also inherit types. I wanted to wait for `src/sentry/api/base.py` to get added to mypy but I think the right thing to is just explicitly type now instead of ignoring.

File: src/sentry/api/endpoints/relay/register_challenge.py (hunk 3 above)
Comment (evanh) [reply src/sentry/api/endpoints/relay/register_challenge.py:57]: My question isn't so much that your typing is wrong. I'm asking more, are there going to be any side effects of returning a 400 here instead of a 500? I don't know what this endpoint is used for.

Comment (untitaker) [reply src/sentry/api/endpoints/relay/register_challenge.py:57]: this is fine, let's do it

File: src/sentry/incidents/serializers/alert_rule.py [hunk 4]
```diff
@@ -243,7 +243,6 @@ def validate(self, data):
             try:
                 raw_snql_query(snql_query, referrer="alertruleserializer.test_query")
             except Exception:
-                # TODO MARCOS 1
```
Comment (untitaker) [inline src/sentry/incidents/serializers/alert_rule.py:246]: hm?

**Review by untitaker — CHANGES_REQUESTED** [review]
_this contains way too many changes at once_. i don't see a problem with any of them but _please_ split up your PRs by owning team such that this can be reviewed effectively by a small amount of people on each PR, not one mega-PR with 10 stakeholders.

**Review by untitaker — APPROVED** [review]

**Comment by untitaker** [comment]
@mgaeta pls coordinate with ingest team via slack (`#discuss-ingest`) so we can watch the deploy, as this might affect services like relay that people don't typically monitor while deploying

File: src/sentry/api/endpoints/relay/register_challenge.py (hunk 3 above)
Comment (jan-auer) [reply src/sentry/api/endpoints/relay/register_challenge.py:57]: I think the intent is to send a `403 Forbidden`  response here, see:
https://github.com/getsentry/relay/blob/2e924639d7bcfa24db69ba2ed78a82e2c07478e1/relay-server/src/actors/upstream.rs#L86-L94

Comment (jan-auer) [reply src/sentry/api/endpoints/relay/register_challenge.py:57]: Double-checked, `400 Bad Request` is fine in this case. It's the same as if we failed to parse JSON.

**Review by jan-auer — APPROVED** [review]
